issue: Manager stop working if a registered Agent cannot be communicated

# Issue report

**Expected Behavior**
PPDDM Manager should continue working with the connected Agents even if there are communication problems with some of them. The problems should be logged and the manager should proceed flawlessly.

**Steps to Reproduce the Problem**

  1. Configure more than 1 Agents for the PPDDM Manager
  2. Make at least one of the configured Agents "un-accessible"
  3. Run the Manager and POST a dataset creation request.

**Logs**

```
akka.stream.StreamTcpException: Tcp command [Connect(localhost:8001,None,List(),Some(10 seconds),true)] failed because of java.net.ConnectException: Connection refused: no further information
Caused by: java.net.ConnectException: Connection refused: no further information
	at java.base/sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:779)
	at akka.io.TcpOutgoingConnection$$anonfun$connecting$1.$anonfun$applyOrElse$3(TcpOutgoingConnection.scala:104)
	at akka.io.TcpOutgoingConnection.akka$io$TcpOutgoingConnection$$reportConnectFailure(TcpOutgoingConnection.scala:51)
	at akka.io.TcpOutgoingConnection$$anonfun$connecting$1.applyOrElse(TcpOutgoingConnection.scala:104)
	at akka.actor.Actor.aroundReceive(Actor.scala:534)
	at akka.actor.Actor.aroundReceive$(Actor.scala:532)
	at akka.io.TcpConnection.aroundReceive(TcpConnection.scala:32)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:573)
	at akka.actor.ActorCell.invoke(ActorCell.scala:543)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:269)
	at akka.dispatch.Mailbox.run(Mailbox.scala:230)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:242)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)

```

--- a/ppddm-agent/src/main/scala/ppddm/agent/controller/prepare/DataPreparationController.scala
+++ b/ppddm-agent/src/main/scala/ppddm/agent/controller/prepare/DataPreparationController.scala
@@ -131,10 +131,13 @@ object DataPreparationController {
           try {
             // Save the dataFrame which includes the prepared data into ppddm-store/datasets/:dataset_id
             DataStoreManager.saveDF(DataStoreManager.getDatasetPath(dataPreparationRequest.dataset_id), dataFrame)
+            logger.info("Prepared data has been successfully saved.")
           }
           catch {
             case e: Exception =>
-              throw DataPreparationException(s"Cannot save the Dataframe of the prepared data with id: ${dataPreparationRequest.dataset_id}.", e)
+              val msg = s"Cannot save the Dataframe of the prepared data with id: ${dataPreparationRequest.dataset_id}."
+              logger.error(msg)
+              throw DataPreparationException(msg, e)
           }
 
           try {
@@ -145,12 +148,15 @@ object DataPreparationController {
               DataStoreManager.saveDF(
                 DataStoreManager.getStatisticsPath(dataPreparationRequest.dataset_id),
                 Seq(dataPreparationResult.toJson).toDF())
+              logger.info("Calculated statistics have been successfully saved.")
             } else {
               logger.warn("This should not have HAPPENED!!! There are no variables in the data preparation request.")
             }
           } catch {
             case e: Exception =>
-              logger.error(s"Cannot save Dataframe with id: ${dataPreparationRequest.dataset_id}.", e)
+              val msg = s"Cannot save Dataframe with id: ${dataPreparationRequest.dataset_id}."
+              logger.error(msg)
+              throw DataPreparationException(msg, e)
           }
 
           // We print the schema and the data only for debugging purposes. Will be removed in the future.
@@ -173,7 +179,7 @@ object DataPreparationController {
     val fields = featureset.variables.get
       .map(variable =>
         StructField(
-          variable.name /*.replaceAll("\\s", "")*/ ,
+          variable.name.trim.replaceAll("\\s", ""),
           if (variable.variable_data_type == VariableDataType.NUMERIC) DoubleType else StringType
         )
       )

--- a/ppddm-manager/src/main/scala/ppddm/manager/controller/dataset/DatasetController.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/controller/dataset/DatasetController.scala
@@ -46,7 +46,7 @@ object DatasetController {
           logger.debug("Inserted document _id:{} and datasetId:{}", _id, datasetWithDataSources.dataset_id.get)
           datasetWithDataSources
         }
-        .recoverWith {
+        .recover {
           case e: Exception =>
             val msg = s"Error while inserting a Dataset with dataset_id:${datasetWithDataSources.dataset_id.get} into the database."
             logger.error(msg, datasetWithDataSources.dataset_id.get, e)

--- a/ppddm-manager/src/main/scala/ppddm/manager/controller/featureset/FeaturesetController.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/controller/featureset/FeaturesetController.scala
@@ -34,7 +34,7 @@ object FeaturesetController {
         logger.debug("Inserted document _id:{} and featuresetId:{}", _id, featuresetWithId.featureset_id.get)
         featuresetWithId
       }
-      .recoverWith {
+      .recover {
         case e: Exception =>
           val msg = s"Error while inserting a Featureset with featureset_id:${featuresetWithId.featureset_id.get} into the database."
           logger.error(msg, featuresetWithId.featureset_id.get, e)

--- a/ppddm-manager/src/main/scala/ppddm/manager/controller/query/FederatedQueryManager.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/controller/query/FederatedQueryManager.scala
@@ -1,8 +1,9 @@
 package ppddm.manager.controller.query
 
 import akka.http.scaladsl.Http
-import akka.http.scaladsl.model.headers.{Accept, Authorization}
 import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.{Accept, Authorization}
+import akka.http.scaladsl.unmarshalling.Unmarshal
 import com.typesafe.scalalogging.Logger
 import ppddm.core.rest.model._
 import ppddm.core.util.URLUtil
@@ -11,7 +12,7 @@ import ppddm.manager.registry.AgentRegistry
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
-import scala.concurrent.duration.{FiniteDuration, MILLISECONDS}
+import scala.util.{Failure, Success, Try}
 
 /* To use the toJson, toPrettyJson methods of the JsonFormatter */
 import ppddm.core.util.JsonFormatter._
@@ -45,20 +46,20 @@ object FederatedQueryManager {
   def invokeAgentsDataPreparation(dataset: Dataset): Future[Dataset] = {
     logger.debug("Will invoke the prepare endpoints of the registered agents")
 
-    Future.sequence(AgentRegistry.dataSources.map { dataSource =>
-      invokeDataPreparation(dataSource, dataset)
-    }) map { responses =>
-      val successfulAgents = responses
-        .filter(_.isDefined) // Filter only the Nonempty responses
-        .map(_.get) // Convert them to Nonoption objects
-      val unsuccessfulAgentSize = AgentRegistry.dataSources.size - successfulAgents.size
-      if (unsuccessfulAgentSize > 0) {
-        logger.warn("There are {} agents (data sources) which returned error on data preparation request.", unsuccessfulAgentSize)
+    Future.sequence(
+      AgentRegistry.dataSources.map { dataSource =>
+        invokeDataPreparation(dataSource, dataset)
       }
-      successfulAgents
-    } map { datasetSources =>
+    ) map { responses =>
+      val failedAgents = responses.collect { case Failure(x) => x }
+      if (failedAgents.nonEmpty) {
+        logger.error("There are {} agents (data sources) out of {} which returned error on data preparation request.", failedAgents.size, responses.size)
+        failedAgents.foreach(logger.error("Error during Agent communication for data preparation", _))
+      }
+
+      val successfulAgents = responses.collect { case Success(x) => x }
       dataset
-        .withDataSources(datasetSources) // create a new Dataset with the DatasetSources which are QUERYING
+        .withDataSources(successfulAgents) // create a new Dataset with the DatasetSources which are QUERYING
         .withExecutionState(ExecutionState.QUERYING) // Set the ExecutionState of the Dataset itself to QUERYING
     }
   }
@@ -70,7 +71,7 @@ object FederatedQueryManager {
    * @param dataset
    * @return
    */
-  private def invokeDataPreparation(dataSource: DataSource, dataset: Dataset): Future[Option[DatasetSource]] = {
+  private def invokeDataPreparation(dataSource: DataSource, dataset: Dataset): Future[Try[DatasetSource]] = {
     val dataPreparationRequest: DataPreparationRequest = DataPreparationRequest(dataset.dataset_id.get, dataset.featureset, dataset.eligibility_criteria, dataset.created_by)
     val uri = URLUtil.append(dataSource.endpoint, DATA_PREPARATION_PATH)
 
@@ -83,18 +84,26 @@ object FederatedQueryManager {
 
     logger.debug("Invoking agent data preparation on URI:{} for dataset_id: {} & dataset_name: {}", uri, dataset.dataset_id, dataset.name)
 
-    Http().singleRequest(request) map {
-      case resp if resp.status == StatusCodes.OK =>
-        logger.debug("Agent data preparation invocation successful on URI:{} for dataset_id: {} & dataset_name: {}", uri, dataset.dataset_id, dataset.name)
-        Some(DatasetSource(dataSource, None, None, Some(ExecutionState.QUERYING)))
-      case errUnk =>
-        errUnk.entity.toStrict(FiniteDuration(1000, MILLISECONDS)).map(_.data.utf8String).map { entity =>
-          logger.error(entity)
-          throw AgentCommunicationException(entity)
+    Http().singleRequest(request).map(Try(_)) flatMap {
+      case Success(res) =>
+        res.status match {
+          case StatusCodes.OK =>
+            logger.debug("Agent data preparation invocation successful on URI:{} for dataset_id: {} & dataset_name: {}", uri, dataset.dataset_id, dataset.name)
+            Future {
+              Success(DatasetSource(dataSource, None, None, Some(ExecutionState.QUERYING)))
+            }
+          case _ =>
+            // I got status code I didn't expect so I wrap it along with body into Future failure
+            Unmarshal(res.entity).to[String].flatMap { body =>
+              throw AgentCommunicationException(dataSource.name, dataSource.endpoint, s"The response status is ${res.status} [${request.uri}] and response body is $body")
+            }
         }
-        None
+      case Failure(e) =>
+        throw e
+    } recover {
+      case e: Exception =>
+        Failure(AgentCommunicationException(dataSource.name, dataSource.endpoint, "Exception while connecting to the Agent",e))
     }
-
   }
 
   def getPreparedDataStatistics(dataset_id: String): Future[Option[DataPreparationResult]] = {

--- a/ppddm-manager/src/main/scala/ppddm/manager/exception/AgentCommunicationException.scala
+++ b/ppddm-manager/src/main/scala/ppddm/manager/exception/AgentCommunicationException.scala
@@ -1,5 +1,5 @@
 package ppddm.manager.exception
 
-final case class AgentCommunicationException(private val reason: String, private val cause: Throwable = None.orNull)
+final case class AgentCommunicationException(name: String, url: String, reason: String, cause: Throwable = None.orNull)
   extends Exception(reason: String, cause: Throwable) {
 }

